minetest/minetest
# Fix rendering regression with TGA type 1 files with BGRA8 color

Irrlicht wildly mixes *byte order* conventions and *endianness order* conventions.

CColorConverter::convert_B8G8R8A8toA8R8G8B8()
would take *byte order* BGRA and convert that into *byte order* ARGB by reversing the order of bytes.

CColorConverter::convert_B8G8R8toA8R8G8B8() on the other hand would take *byte order* BGR and convert that into *endianness order* ARGB.

TGA uses BGR(A)8 *byte order*. Irrlicht typically expects *endianness order* ARGB8 (see SColor, CImage etc.; though inconsistently, e.g. CImage::getPixel uses *byte order* for RGB8).

This makes both conversion functions *byte order* to *endianness order* ARGB8 as they should have been, documents them properly and moves them to CImageLoaderTGA.cpp so no poor soul shall be fooled by them in the near future.

Based on an original patch by erle: http://daten.dieweltistgarnichtso.net/src/0001-Fix-regressions-in-rendering-TGA-type-1-files-with-c.patch which I found very confusing, because their implementation of the BGRA8 -> ARGB8 conversion is essentially just a copy on little endian, which led me to discover (and attempt to clean up a *slight bit*, just enough to resolve this issue, and sprinkle in a couple `// FIXME`s so this doesn't screw anyone over too badly in the future) this hellhole of Irrlicht inconsistencies.

## How to test

1. As with https://github.com/minetest/minetest/pull/15390, get `tga_testnodes`.
2. Place all nodes, or at least the type 1 BGR(A)8 ones.
3. BGR8 ones should display red, green, blue and a black background, BGRA8 ones should have a semitransparent background instead.

It should look like this: ![image](https://github.com/user-attachments/assets/a42f4220-88be-439f-a590-f211010f703a)

It should *not* look something like this: ![image](https://github.com/user-attachments/assets/d283374d-96f4-486e-b238-e2ee281000e0)
